src/node.cpp

```diff
@@ -133,24 +101,48 @@ NodeSession& Node::getNodeSession(const session_key_t& sessionKey) const {
 }
 
 NodeSession& Node::getNodeSession(const NodeSessionMetadata& metadata) {
-    auto it = nodeSessions.find(metadata.getSessionKey());
+    session_key_t sessionKey;
+    if (gatherFrom) {
+        sessionKey = metadata.getSessionKey(gatherFrom.value());
+    } else {
+        sessionKey = metadata.getSessionKey();
+    }
+    auto it = nodeSessions.find(sessionKey);
     if (it != nodeSessions.end()) {
         return *(*it).second;
     }
     SPDLOG_LOGGER_DEBUG(dag_executor_logger, "Will create new session: {} for node: {}",
         metadata.getSessionKey(), getName());
-    auto emplacePair = nodeSessions.emplace(metadata.getSessionKey(), createNodeSession(metadata));
+    std::optional<uint16_t> expectedEachInputsCount;
+    if (gatherFrom) {
+        expectedEachInputsCount = 0;
+        for (auto nodeName : gatherFrom.value()) {
+            expectedEachInputsCount.value() += metadata.getSubsessionSize(nodeName);
+        }
+    }
+    NodeSessionMetadata newSessionMetadata;
+    uint16_t shardsCount = 1;
+    if (gatherFrom) {
+        for (auto& nodeName : gatherFrom.value()) {
+            shardsCount *= metadata.getSubsessionSize(nodeName);
+        }
+        newSessionMetadata = metadata.getCollapsedSessionMetadata(gatherFrom.value());
+    } else {
+        newSessionMetadata = metadata;
+    }
+    std::unique_ptr<NodeSession> nodeSession = createNodeSession(newSessionMetadata, shardsCount);
+    auto emplacePair = nodeSessions.emplace(sessionKey, std::move(nodeSession));
     return *(*(emplacePair.first)).second;
```

Comment:
You can reduce nesting and parentheses by using `->` operator.
`emplacePair->first->second`